src/google-classroom-extractor/google_classroom_extractor/api/courses.py

```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
+        # select unmatched records into temp table - differing by hash for same id
+        # single entry in result set if id only exists in one table (meaning add or missing),
+        #   so SyncNeeded flag will indicate which table it's from
+        # double entry in result set if id exists in both (meaning update needed),
+        #   so SyncNeeded will show which row is from which table
+        con.execute(
+            "CREATE TABLE Unmatched AS "
+            "SELECT * FROM ( "
+            "    SELECT * FROM Courses "
+            "    UNION ALL "
+            "    SELECT * FROM Sync_Courses "
+            ") "
+            "GROUP BY id, Hash "
+            "HAVING COUNT(*) = 1"
+        )
+        con.execute("CREATE INDEX ID_INDEX ON Unmatched(id)")
+
+        # all rows start with CreateDate and LastModifiedDate initialized to "now",
+        #   but updated rows need the original CreateDate pulled from existing table
+        # Note: UPDATE-FROM is not available in sqlite until v3.33.0, thus the
+        #   double select goofiness
+        con.execute(
+            "UPDATE Unmatched "
+            "    SET CreateDate = ( "
+            "        SELECT c.CreateDate "
+            "        FROM Courses c "
+            "        WHERE c.id = Unmatched.id "
+            "    ) "
+            "    WHERE EXISTS ( "
+            "        SELECT * "
+            "        FROM Courses c "
+            "        WHERE c.id = Unmatched.id "
+            "    ) AND SyncNeeded = 1"
+        )
+
+        # fetch DataFrame with the changed record CreateDate updates
+        create_date_query = ("SELECT id, CreateDate "
+                             "FROM Unmatched "
+                             "WHERE id IN ( "
+                             "    SELECT id FROM Unmatched "
+                             "    GROUP BY id "
+                             "    HAVING COUNT(*) > 1 "
+                             ") AND SyncNeeded = 1")
+        create_date_df = read_sql_query(create_date_query, con)
+        create_date_df.set_index('id', inplace=True)
+
+        # update the CreateDates of updated rows
+        courses_df.update(create_date_df)
+
+        # now do deletes and inserts
+        con.execute(
+            "WITH "
+            #    changed rows CTE (from SyncNeeded side only)
+            "    changedRows AS ( "
+            "        SELECT * FROM Unmatched "
+            "        WHERE id IN ( "
+            "            SELECT id FROM Unmatched "
+            "            GROUP BY id "
+            "            HAVING COUNT(*) > 1 "
+            "        ) AND SyncNeeded = 1 "
+            "    )"
+
+
+            # delete the obsolete data
+            "DELETE FROM Courses "
+            "WHERE id IN ( "
+            "    SELECT id from changedRows "
+            ")"
+        )
+
+        con.execute(
+            # insert new and changed data
+            "WITH "
+            #    changed rows CTE (from SyncNeeded side only)
```

Comment:
Right now duplicating this CTE to select only the change rows from Unmatched table. Need to either restructure or figure out how to share across executions with SQlAlchemy (didn't need to do this in script)